lib/phoenix/presence.ex

```diff
@@ -343,29 +343,203 @@ defmodule Phoenix.Presence do
       pubsub_server =
         opts[:pubsub_server] || raise "use Phoenix.Presence expects :pubsub_server to be given"
 
-      Phoenix.Tracker.start_link(__MODULE__, {module, task_supervisor, pubsub_server}, opts)
+      presence_client = opts[:presence_client]
+
+      Phoenix.Tracker.start_link(
+        __MODULE__,
+        {module, task_supervisor, pubsub_server, presence_client},
+        opts
+      )
     end
 
     def init(state) do
       {:ok, state}
     end
 
     def handle_diff(diff, state) do
-      {module, task_supervisor, pubsub_server} = state
+      {module, task_supervisor, pubsub_server, presence_client} = state
 
       Task.Supervisor.start_child(task_supervisor, fn ->
         for {topic, {joins, leaves}} <- diff do
-          Phoenix.Channel.Server.local_broadcast(pubsub_server, topic, "presence_diff", %{
+          presence_diff = %{
             joins: module.fetch(topic, Phoenix.Presence.group(joins)),
             leaves: module.fetch(topic, Phoenix.Presence.group(leaves))
-          })
+          }
+
+          if presence_client do
+            Phoenix.Presence.Client.handle_diff(topic, presence_diff)
+          end
+
+          Phoenix.Channel.Server.local_broadcast(
+            pubsub_server,
+            topic,
+            "presence_diff",
+            presence_diff
+          )
         end
       end)
 
       {:ok, state}
     end
   end
 
+  defmodule Client do
```

Comment:
Rather than an extra process here, we can put everything that this genserver is doing inside of the `Phoenix.Presence` module. For example, `init` here will move up inside `Phoenix.Presence.init`, and then we can handle the state inside the same process without extra copying